### PR TITLE
#389 Prevent progressbar spill over new line

### LINKED ISSUE #389: Incorrect progress bar display
Sometimes the progressbar goes bad and display each message on a new line
It could be because the file name displayed exceed the terminal width and spills over the next line, detroying the progressbar curses sync along the way.
:
```
[-----------------------------------#]  95981  Scanned: c                                                                                                       [-----------------------------------#]  95982  Scanned: f                                                                                                       [-----------------------------------#]  95983  Scanned: 2                                                                                                       [-----------------------------------#]  95984  Scanned: 6                                                                                                       [-----------------------------------#]  95985  Scanned: b                                                                                                       [-----------------------------------#]  95986  Scanned: build                                                                                                   [-----------------------------------#]  95987  Scanned: release                                                                                                 [-----------------------------------#]  95988  Scanned: conf                                                                                                    [-----------------------------------#]  95989  Scanned: scripts                                                                                                 [-----------------------------------#]  95990  Scanned: MANIFEST.in.dev                                                                                         [-----------------------------------#]  95991  Scanned: configure.py                                                                                            [-----------------------------------#]  95992  Scanned: scancode-toolkit.spec                                                                                   [-----------------------------------#]  95993  Scanned: release.sh                                                                                              [-----------------------------------#]  95994  Scanned: setup.cfg.dev                                                                                           [-----------------------------------#]  95995  Scanned: MANIFEST.in.release                                                                                     [-----------------------------------#]  95996  Scanned: dev                                                                                                     [-----------------------------------#]  95997  Scanned: base.txt                                                                                                [-----------------------------------#]  95998  Scanned: base.py                                                                                                 [-----------------------------------#]  95999  Scanned: base.txt                                                                                                [-----------------------------------#]  96000  Scanned: irc-notify.py.ABOUT                                                                                     [-----------------------------------#]  96001  Scanned: test_json2csv.py                                                                                        [-----------------------------------#]  96002  Scanned: irc-notify.py                                                                                           [-----------------------------------#]  96003  Scanned: README.rst                                                                                              [-----------------------------------#]  96004  Scanned: dejacode.py                                                                                             [-----------------------------------#]  96005  Scanned: json2csv.py                                                                                             [-----------------------------------#]  96006  Scanned: testdata                                                                                                [-----------------------------------#]  96007  Scanned: json2csv                                                                                                [-----------------------------------#]  96008  Scanned: no_keys.json                                                                                            [-----------------------------------#]  96009  Scanned: minimal.json                                                                                            [-----------------------------------#]  96010  Scanned: gpl-2.0.LICENSE                                                                                         [-----------------------------------#]  96011  Scanned: minimal.csv                                                                                             [-----------------------------------#]  96012  Scanned: setup.cfg.release                                                                                       [-----------------------------------#]  96013  Scanned: key_order.json                                                                                          [-----------------------------------#]  96014  Scanned: minimal.json-expected                                                                                   [-----------------------------------#]  96015  Scanned: minimal.json-strip-expected                                                                             [-----------------------------------#]  96016  Scanned: format_html_app_data.json                                                                               [-----------------------------------#]  96017  Scanned: format_json_scan.json                                                                                   [-----------------------------------#]  96018  Scanned: full.json                                                                                               [-----------------------------------#]  96019  Scanned: full.json-expected                                                                                      [####################################]  96019                        
```

### PR DESCRIPTION
**Pull request by yasharmaster**
Fixes #389 
Signed-off-by: Yash Sharma yasharmaster@gmail.com

### CODE REVIEW

**Comment by yasharmaster** [comment]
@pombredanne I have hard coded the allowed size of a filename to be 20 characters.
Please review the changes.

File: src/scancode/cli.py [hunk 1]
```diff
@@ -382,6 +382,9 @@ def scan_event(item):
                     return ''
                 if item:
                     _scan_success, _scanned_path = item
+                    if len(_scanned_path)>20:
```
Comment (pombredanne) [inline src/scancode/cli.py]: Your spacing is not right. We use PEP8 for code standard. eg. `len(scanned_path) > 20`

File: src/scancode/cli.py [hunk 2]
```diff
@@ -382,6 +382,9 @@ def scan_event(item):
...
+                    	current_filename, current_file_extension = os.path.splitext(_scanned_path)
```
Comment (pombredanne) [inline src/scancode/cli.py]: Please consider the use of the `commoncode.fileutils` functions instead of splitext

File: src/scancode/cli.py [hunk 3]
```diff
@@ -382,6 +382,9 @@ def scan_event(item):
...
+                    	_scanned_path = current_filename[:3] + '...' + current_filename[-3:] + current_file_extension
```
Comment (pombredanne) [inline src/scancode/cli.py]: keeping only three characters at start and end does not sound right.
IMHO you should also isolate this in its own function so it can be tested.

**Comment by yasharmaster** [comment]
@pombredanne please review the changes. I have -
[*] used `commoncode.fileutils` instead of `splitext`
[*] implemented file name shortening in a separate function
[*] set max filename size to be 25 chars, more than 25 was spilling progress bar in my system
[*] improved shortened filename by allowing more than 3 chars from beginning and the end

File: src/scancode/cli.py [hunk 4]
```diff
@@ -443,6 +446,20 @@ def scan_event(item):
     return files_count, cached_scan.iterate(scans, root_dir)
 
 
+def _shorten_file_path(file_path, max_length):
```
Comment (pombredanne) [inline src/scancode/cli.py]: do not use a _ prefix here.

**Comment by pombredanne** [comment]
ping.

**Comment by yasharmaster** [comment]
@pombred

**Comment by yasharmaster** [comment]
I am travelling right now. Will finish these edits by tonight.

**Comment by yasharmaster** [comment]
@pombredanne Default terminal width is 80 chars across all OS and the text before the filename spans across 55 characters. So, only 25 chars can be safely allowed for the filename if we hardcode the value. Otherwise we can do something like this https://gist.github.com/jtriley/1108174 to find current terminal width. What do you think?

**Comment by yasharmaster** [comment]
@pombredanne please review

File: src/scancode/cli.py [hunk 5]
```diff
@@ -44,6 +44,8 @@
 import click
 from click.termui import style
 
+from math import floor
```
Comment (pombredanne) [inline src/scancode/cli.py]: This is a stdlib import and should go with other stdlib imports, before `import os`

File: src/scancode/cli.py [hunk 6]
```diff
@@ -382,6 +384,7 @@ def scan_event(item):
                     return ''
                 if item:
                     _scan_success, _scanned_path = item
+                    _scanned_path = shorten_filename(_scanned_path)
```
Comment (pombredanne) [inline src/scancode/cli.py]: this should only be done if we are not verbose. Note also that you are never passing a filename but a path. So your assumptions are not right in your function.

File: src/scancode/cli.py [hunk 7]
```diff
@@ -443,6 +446,26 @@ def scan_event(item):
     return files_count, cached_scan.iterate(scans, root_dir)
 
 
+def shorten_filename(filename, max_length=25):
+	"""
+	Return the original filename if it is short enough
+	otherwise return its shortened version 
+	"""
+	if len(filename) <= max_length:
+		shortened_filename = filename
```
Comment (pombredanne) [inline src/scancode/cli.py]: why not just return early?

File: src/scancode/cli.py [hunk 8]
```diff
@@ -443,6 +446,26 @@ def scan_event(item):
...
+	else:
```
Comment (pombredanne) [inline src/scancode/cli.py]: if you return early above, `else` is no longer needed

File: src/scancode/cli.py [hunk 9]
```diff
@@ -443,6 +446,26 @@ def scan_event(item):
...
+		file_base_name = fileutils.file_base_name(filename)
+		file_extension = fileutils.file_extension(filename)
+		number_of_dots = 3
+		remaining_length = max_length - len(file_extension) - number_of_dots
+		number_of_beginning_chars = int(floor(remaining_length/2))
```
Comment (pombredanne) [inline src/scancode/cli.py]: pep8 spacing please. using abs and  a floor `//` division would  be simpler and clearer and more explicit:
`abs(remaining_length // 2)`

Also I am not sure that `number_of_beginning_chars` is a good variable name as you use it to compute both the prefix and suffix length to keep from the base name

File: src/scancode/cli.py [hunk 10]
```diff
@@ -382,6 +384,7 @@ def scan_event(item):
                     return ''
                 if item:
                     _scan_success, _scanned_path = item
+                    _scanned_path = shorten_filename(_scanned_path)
                     _progress_line = verbose and _scanned_path or fileutils.file_name(_scanned_path)
```
Comment (pombredanne) [inline src/scancode/cli.py]: This code cannot work  correctly with the new code that you injected?

File: src/scancode/cli.py [hunk 11]
```diff
@@ -443,6 +446,26 @@ def scan_event(item):
     return files_count, cached_scan.iterate(scans, root_dir)
 
 
+def shorten_filename(filename, max_length=25):
+	"""
+	Return the original filename if it is short enough
```
Comment (pombredanne) [inline src/scancode/cli.py]: You are receiving a path and not a name. So your docstring is misleading

File: src/scancode/cli.py [hunk 12]
```diff
@@ -443,6 +446,26 @@ def scan_event(item):
...
+	otherwise return its shortened version 
+	"""
+	if len(filename) <= max_length:
+		shortened_filename = filename
+	else:
+		file_base_name = fileutils.file_base_name(filename)
+		file_extension = fileutils.file_extension(filename)
```
Comment (pombredanne) [inline src/scancode/cli.py]: use instead  something like 
`base_name, ext = fileutils.splitext(filename)` .... but again you did not receive a filename but a path (though it does not matter to splitext)

File: src/scancode/cli.py [hunk 13]
```diff
@@ -443,6 +446,26 @@ def scan_event(item):
...
+		number_of_dots = 3
+		remaining_length = max_length - len(file_extension) - number_of_dots
+		number_of_beginning_chars = int(floor(remaining_length/2))
+		shortened_filename = (file_base_name[:number_of_beginning_chars] + 
+							number_of_dots*'.' + 
+							file_base_name[-number_of_beginning_chars:] +
```
Comment (pombredanne) [inline src/scancode/cli.py]: Your pep8 spacing is not correct in this code block

File: tests/scancode/data/long_file_name/0123456789012345678901234567890123456789.c [hunk 14]
```diff
@@ -0,0 +1,1967 @@
+/* deflate.c -- compress data using the deflation algorithm
```
Comment (pombredanne) [inline tests/scancode/data/long_file_name/0123456789012345678901234567890123456789.c]: Why use such a big file for tests when you only care about the name? and not the content? This is wasted space.

File: tests/scancode/data/long_file_name/expected.json [hunk 15]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (pombredanne) [inline tests/scancode/data/long_file_name/expected.json]: You should not care about the detailed scan results when you tests name shortening in the CLI ui.

File: tests/scancode/test_cli.py [hunk 16]
```diff
@@ -169,6 +169,15 @@ def test_scan_info_does_collect_infos_with_root(monkeypatch):
...
 
+def test_scan_can_handle_long_file_names(monkeypatch):
+    monkeypatch.setattr(click._termui_impl, 'isatty', lambda _: True)
+    test_dir = test_env.get_test_loc('long_file_name/0123456789012345678901234567890123456789.c')
+    runner = CliRunner()
+    result_file = test_env.get_temp_file('json')
+    result = runner.invoke(cli.scancode, ['--copyright', test_dir, result_file], catch_exceptions=True)
+    assert result.exit_code == 0
+    assert 'Scanning done' in result.output
+    check_scan(test_env.get_test_loc('long_file_name/expected.json'), result_file, regen=False)
```
Comment (pombredanne) [inline tests/scancode/test_cli.py]: This is not needed. We do not care about the scan. On the other hand you make no assertions on the shortening of the name in the UI, so you are not testing anything there

File: src/scancode/cli.py (hunk 13 above)
Comment (pombredanne) [inline src/scancode/cli.py]: You should prefer a more explicit interpolation or format. Something like `'{start}...{end}'.format(**locals())` were you compute `start` and `end` before.

**Comment by pombredanne** [comment]
@yasharmaster When you are done with updating your code, you will also need to squash your commit in one clean commit. But only then.

**Comment by yasharmaster** [comment]
@pombredanne Please review. Should I send a new PR with this last single commit ?

**Comment by pombredanne** [comment]
@yasharmaster nope, you can just rebase, squash and force push to that same branch of yours. This will land right here in the same PR aqlright.

**Comment by yasharmaster** [comment]
@pombredanne Can you please point me towards the existing unit tests in the repo? Also the location where unit tests of shorten filename should go? Thanks

**Comment by pombredanne** [comment]
Just look anywhere in the tests/ directory. There are like 1000's of them. You can add your tests alright in the same test_cli.py module
An example of a simple unit tests could be this: https://github.com/nexB/scancode-toolkit/blob/21b6c8456f6cfd42ffc07411312cc489adec26d5/tests/commoncode/test_fileutils.py#L297

**Comment by yasharmaster** [comment]
@pombredanne Are these unit tests sufficient? I have now addressed all the comments from your previous review.

File: tests/scancode/test_cli.py [hunk 17]
```diff
@@ -463,3 +464,17 @@ def test_scan_can_run_from_other_directory():
         print(stderr)
     assert rc == 0
     check_scan(test_env.get_test_loc(expected_file), result_file, strip_dates=True, regen=False)
+
+
+class TestCLI(TestCase):
+
+	def test_shorten_filename(self):
```
Comment (pombredanne) [inline tests/scancode/test_cli.py]: You should use 4 spaces. not 8

File: tests/scancode/test_cli.py [hunk 18]
```diff
@@ -463,3 +464,17 @@ def test_scan_can_run_from_other_directory():
...
+	    assert '0123456789...0123456789.c' == cli.shorten_filename('0123456789012345678901234567890123456789.c')
+	    assert '012345.c' == cli.shorten_filename('012345.c')
+	    assert '012345678901234567890.cpp' == cli.shorten_filename('012345678901234567890.cpp')
```
Comment (pombredanne) [inline tests/scancode/test_cli.py]: you should add tests for limit args: empty, None, list, string of a length that is and a length that is not shortened right around your 25 limit.  It may help to use variables and give names to make things more explicit. Also you need to test things without extensions and full paths.
In general it is better to create one test function for one test, though the example I gave you did not follow that pattern.

File: src/scancode/cli.py [hunk 19]
```diff
@@ -383,6 +384,8 @@ def scan_event(item):
                 if item:
                     _scan_success, _scanned_path = item
                     _progress_line = verbose and _scanned_path or fileutils.file_name(_scanned_path)
+                    if not verbose:
+                    	_progress_line = shorten_filename(_progress_line)
```
Comment (yasharmaster) [inline src/scancode/cli.py]: @pombredanne The logic here is such that `_progress_line` will contain the file path only if `verbose` is set to True. In that case `shorten_filename()` will not be called at all. So argument passed to  `shorten_filename()` will always be filename only and not the file path. So why are we testing `shorten_filename()` with a file path?

File: tests/scancode/test_cli.py (hunk 18 above)
Comment (yasharmaster) [inline tests/scancode/test_cli.py]: @pombredanne what should be the output of shorten filename if a list is passed as the filename?

File: src/scancode/cli.py (hunk 19 above)
Comment (pombredanne) [inline src/scancode/cli.py]: good point.

File: tests/scancode/test_cli.py (hunk 18 above)
Comment (pombredanne) [inline tests/scancode/test_cli.py]: what do you think? probably an exception

**Comment by yasharmaster** [comment]
@pombredanne I have merged all commits into one!

File: tests/scancode/test_cli.py [hunk 20]
```diff
@@ -104,6 +105,17 @@ def test_verbose_option_with_copyrights(monkeypatch):
...
 
+def test_scan_can_handle_long_file_names(monkeypatch):
+    monkeypatch.setattr(click._termui_impl, 'isatty', lambda _: True)
+    test_dir = test_env.get_test_loc('long_file_name/0123456789012345678901234567890123456789.c')
+    runner = CliRunner()
+    result_file = test_env.get_temp_file('json')
+    result = runner.invoke(cli.scancode, ['--copyright', test_dir, result_file], catch_exceptions=True)
+    assert result.exit_code == 0
+    assert 'Scanning done' in result.output
+    assert '0123456789...0123456789.c' in result.output
```
Comment (pombredanne) [inline tests/scancode/test_cli.py]: Perfect. You are getting it.

File: tests/scancode/test_cli.py [hunk 21]
```diff
@@ -452,3 +464,41 @@ def test_scan_can_run_from_other_directory():
...
     check_scan(test_env.get_test_loc(expected_file), result_file, strip_dates=True, regen=False)
+
+
+class TestShortenFilename(TestCase):
+
+    def test_filename_is_shortened(self):
+        test = cli.shorten_filename(filename='012345678901234567890123.c')
+        expected = '0123456789...4567890123.c'
+        assert expected == test
+
```
Comment (pombredanne) [inline tests/scancode/test_cli.py]: Use only one blank line between class methods. Two lines is only between top level module members

File: tests/scancode/test_cli.py [hunk 22]
```diff
@@ -452,3 +464,41 @@ def test_scan_can_run_from_other_directory():
...
+
+    def test_without_extension(self):
+        test = cli.shorten_filename(filename='012345678901234567890123456')
+        expected = '01234567890...67890123456'
+        assert expected == test
+
+
+    def test_with_full_path(self):
```
Comment (pombredanne) [inline tests/scancode/test_cli.py]: Good test. You should also have a test with a non-posix path

Comment (pombredanne) [inline tests/scancode/test_cli.py]: And tests with our without posix paths that do shortening or not

File: tests/scancode/test_cli.py [hunk 23]
```diff
@@ -452,3 +464,41 @@ def test_scan_can_run_from_other_directory():
         print(stderr)
     assert rc == 0
     check_scan(test_env.get_test_loc(expected_file), result_file, strip_dates=True, regen=False)
+
+
+class TestShortenFilename(TestCase):
```
Comment (pombredanne) [inline tests/scancode/test_cli.py]: You should add tests that test at the boundaries: e.g. when you have exactly the max length that will be shorted., then that minus one and that plus one. This way this would spec your function fully. You may have them already, but if this is the case, the test function names are not explicit enough

File: src/scancode/cli.py [hunk 24]
```diff
@@ -467,6 +469,26 @@ def scan_event(item):
     return files_count, cached_scan.iterate(scans, root_dir)
 
 
+def shorten_filename(filename='', max_length=25, number_of_dots=3):
+    """
+    Return the original filename if it is short enough
```
Comment (pombredanne) [inline src/scancode/cli.py]: What about rather:
`Return a truncated `filename` usable for fixed width display if `filename` is longer than `max_length`.`
Also why do we need a default for `filename=''`? Calling this function without args doe not make sense. It should just error out. Nothing special needed. `number of dots` should not be an arg. This is not something I would need to ever modify.

Comment (yasharmaster) [inline src/scancode/cli.py]: @pombredanne If we don't keep `filename=''` as default, test with empty args will fail! Should I remove the test with empty args?

Comment (pombredanne) [inline src/scancode/cli.py]: Or you can test ass an exercise that the proper exception is raised in this case.

**Comment by yasharmaster** [comment]
@pombredanne Since `shorten_filename()` will not be called without args, therefore I have removed the test with no args. I have added new tests and have renamed them to be more explicit.

**Comment by pombredanne** [comment]
@yasharmaster you wrote in #aboutcode IRC:
> Could anyone please let me know if should raise an exception if the function is called with no args or let it just error out? 

Just let it error out alright. My only point is that if you want to have a comprehensive coverage, then you could test that an exception is thrown in these cases. Not really needed in general but a possible useful excercise for you

**Comment by yasharmaster** [comment]
@pombredanne please review

**Comment by pombredanne** [comment]
This looks good to me. I am merging soon.

**Comment by pombredanne** [comment]
as a side note, I wished we would have something based on the TTY width, such as with `blessings` instead of computing things based on a guessed fixed width.

**Comment by yasharmaster** [comment]
@pombredanne Since the code has been merged after rebase, this can be closed as well.

**Comment by pombredanne** [comment]
@yasharmaster closing as this was merged. Thanks++
